Fix ResultSink and it's unit tests

david415 opened the pull request:
fixed many bugs and implementation mistakes with ResultSink. Please review.

david415 committed:
Minor fix to random_path

DonnchaC commented on test/template.py:
```diff
@@ -34,10 +34,12 @@ def exits(self):
-        return selected_relays[0:2] + [exit_relay]
+        non_exits = [r for r in self.routers if 'exit' not in r.flags]
+        if len(non_exits) > 1:
+            return non_exits[0:2] + [exit_relay]
```
I don't see a reason to only select the first two relays from the set of non-exits. Shouldn't those relays be selected from the full set of relays?

DonnchaC commented on test/template.py:
```diff
@@ -34,10 +34,12 @@ def exits(self):
+        if len(non_exits) > 1:
+            return non_exits[0:2] + [exit_relay]
+        else:
+            selected_relays = random.sample(self.routers, 3)
```
You might chose a non-exit relay as the final hop here. This will cause random breakage in the test?

david415 committed:
Fix bug in writer and tests

make send always return a deferred

david415 committed:
Clean writer end_flush

david415 committed:
Teach writer send to always flush all the whole chunks from buffer

- here we also teach end_flush to add it's task to the deferred chain
and execute it in the current thread

david415 committed:
Attempt to fix ResultSink unit test class's tearDown method

david415 committed:
Revert "Minor fix to random_path"

This reverts commit 3f4f3efc391e401de7d573def8adf899a502f42a.
